llvm/lib/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.cpp

```diff
@@ -10,10 +10,69 @@
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/SandboxIR/Function.h"
 #include "llvm/SandboxIR/Instruction.h"
-
-using namespace llvm::sandboxir;
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Transforms/Vectorize/SandboxVectorizer/Passes/NullPass.h"
 
 namespace llvm::sandboxir {
+
+static cl::opt<bool>
+    PrintPassPipeline("sbvec-print-pass-pipeline", cl::init(false), cl::Hidden,
+                      cl::desc("Prints the pass pipeline and returns."));
+
+/// A magic string for the default pass pipeline.
+static const char *DefaultPipelineMagicStr = "*";
+
+static cl::opt<std::string> UserDefinedPassPipeline(
+    "sbvec-passes", cl::init(DefaultPipelineMagicStr), cl::Hidden,
+    cl::desc("Comma-separated list of vectorizer passes. If not set "
+             "we run the predefined pipeline."));
+
+static std::unique_ptr<RegionPass> createRegionPass(StringRef Name) {
+#define REGION_PASS(NAME, CREATE_PASS)                                         \
+  if (Name == NAME)                                                            \
+    return std::make_unique<decltype(CREATE_PASS)>(CREATE_PASS);
+#include "PassRegistry.def"
+  return nullptr;
+}
+
+/// Adds to `RPM` a sequence of region passes described by `Pipeline`.
+static void parseAndCreatePassPipeline(RegionPassManager &RPM,
```

Comment:
Now that it's the pass managers that own the passes it looks a bit out of place that populating them is done with a static function, because the only users of it are the pass managers.

Wdyt about moving this to the PassManager base class?

Reply:
If we do this, then sandboxir::PassManager has to know about vectorizer-specific passes. Is this a desired outcome?

Reply:
I meant making `parseAndCreatePassPipeline` a member function of `PassManager` such that we call:
```
RPM.parseAndCreatePassPipeline(UserDefinedPassPipeline);
```
instead of:
```
parseAndCreatePassPipeline(RPM, UserDefinedPassPipeline);
```

Reply:
Right. But `parseAndCreatePassPipeline` needs to know about the existing vectorizer passes in order to instantiate them.

Do you want llvm/lib/SandboxIR/PassManager.cpp to include all the passes from llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Passes/* ? Should we move `RegionPassManager` back into the vectorizer?

Reply:
Well, conceptually parsing the pass pipeline string and building passes should not be a static function in BottomUpVec.cpp because it is a pass-manager task, not a vectorization-specific task.

Oh yes, the creator function would need to have access to the passes. Hmm could we perhaps keep the creator function here in a lambda `createPass(Name)` and pass the lambda as an argument to the parse function?

Reply:
Done. Left `createPass` as a static function in BottomUpVec.h instead of a lambda because the includes and defines needed for the .def file messed up indentation and I preferred to keep that ugliness contained in its own function.

I also added a new unit test, similar to the now deleted PassRegistry test, that checks the pipeline parsing function (we also have some very basic coverage from lit tests)